src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,315 @@
+        ]
+        |> String.concat "\n"
+
+    static member AddInstrumentation(cacheId) =
+        if observedCaches.ContainsKey cacheId then
+            invalidArg "cacheId" $"cache with name {cacheId} already exists"
+
+        observedCaches[cacheId] <- new CacheMetrics(cacheId)
+
+    static member RemoveInstrumentation(cacheId) =
+        observedCaches[cacheId].Dispose()
+        observedCaches.TryRemove(cacheId) |> ignore
+
+type EntityPool<'Key, 'Value>(totalCapacity, cacheId) =
+    let pool = ConcurrentBag<CachedEntity<'Key, 'Value>>()
+    let mutable created = 0
+
+    let overCapacity =
+        CacheMetrics.Meter.CreateCounter<int64>("over-capacity", "count", cacheId)
+
+    member _.Acquire(key, value) =
+        match pool.TryTake() with
+        | true, entity -> entity.ReUse(key, value)
+        | _ ->
+            if Interlocked.Increment &created > totalCapacity then
+                overCapacity.Add 1L
+
+            CachedEntity(key, value).WithNode()
+
+    member _.Reclaim(entity: CachedEntity<'Key, 'Value>) =
+        if pool.Count < totalCapacity then
+            pool.Add(entity)
+
+module Cache =
+    // During testing a lot of compilations are started in app domains and subprocesses.
+    // This is a reliable way to pass the override to all of them.
+    [<Literal>]
+    let private overrideVariable = "FSHARP_CACHE_OVERRIDE"
+
+    /// Use for testing purposes to reduce memory consumption in testhost and its subprocesses.
+    let OverrideCapacityForTesting () =
+        Environment.SetEnvironmentVariable(overrideVariable, "true", EnvironmentVariableTarget.Process)
+
+    let applyOverride (capacity: int) =
+        match Environment.GetEnvironmentVariable(overrideVariable) with
+        | NonNull _ when capacity > 1024 -> 1024
+        | _ -> capacity
+
+[<Sealed; NoComparison; NoEquality>]
+[<DebuggerDisplay("{GetStats()}")>]
+type Cache<'Key, 'Value when 'Key: not null and 'Key: equality>
+    internal (totalCapacity, headroom, cts: CancellationTokenSource, ?name, ?observeMetrics) =
+
+    let instanceId = defaultArg name (Guid.NewGuid().ToString())
+
+    let observeMetrics = defaultArg observeMetrics false
+
+    do
+        if observeMetrics then
+            CacheMetrics.AddInstrumentation instanceId
+
+    let meter = CacheMetrics.Meter
+    let hits = meter.CreateCounter<int64>("hits", "count", instanceId)
+    let misses = meter.CreateCounter<int64>("misses", "count", instanceId)
+    let evictions = meter.CreateCounter<int64>("evictions", "count", instanceId)
+
+    let evictionFails =
+        meter.CreateCounter<int64>("eviction-fails", "count", instanceId)
+
+    let pool = EntityPool<'Key, 'Value>(totalCapacity, instanceId)
+
+    let store =
+        ConcurrentDictionary<'Key, CachedEntity<'Key, 'Value>>(Environment.ProcessorCount, totalCapacity)
+
+    let evictionQueue = LinkedList<CachedEntity<'Key, 'Value>>()
+
+    let addToEvictionQueue (entity: CachedEntity<'Key, 'Value>) =
```

Comment:
See the eviction maintenance - `addToEvictionQueue` , `updateEvictionQueue`, `tryEvictOne`, `backgroundEviction` - did you consider any tradeoffs between manual management like done here vs. using a `MailboxProcessor` that would encapsulate access to the linked list and provide messages for `add` , `update`?

Reply:
Yes! I considered exactly this. I decided against only because we now do LRU only, so the updates are fast. I'm not sure but I feel this would add overhead.
If we decide for other eviction algorithms, it may be beneficial to queue the queue updates, too. I'll add a comment.

Reply:
On a second thought, a mailbox processor would replace the background eviction task and make the code clearer. I'll try it.